File: lib/JS/PyQuery.js
/*
	If you'd like to make changes to the api, start here.
	This is where you can add handlers for commands and requests
*/

var pyDir = '../lib/PY/main.py';
var root;
var platform;

$(function () {
	var indexedElements = [];
	var placeholderIds = 0;

	python = makePython();
	
	//commands
	python.add.commandHandler('warn', function (data) {
		console.warn('Python Warns:' + data.msg);
	});
	python.add.commandHandler('alert', function (data) {
		alert(data.msg);
	});
	python.add.commandHandler('killMe', function (data) {
		py.kill();
		console.warn("Python is requesting death");
		alert("Python Suicided");
	});
	python.add.commandHandler('css', function (data) {
		getJqueryObj(data.selector, data.index).css(data.rules);
	});
	python.add.commandHandler('html', function (data) {
		getJqueryObj(data.selector, data.index).html(data.text);
	});
	python.add.commandHandler('append', function (data) {
		getJqueryObj(data.selector, data.index).append(data.text);
	});
	python.add.commandHandler('attribute', function (data) {
		getJqueryObj(data.selector, data.index).attr(data.property, data.value);
	});
	python.add.commandHandler('value', function (data) {
		return getJqueryObj(data.selector, data.index).value(data.value);
	});
	python.add.commandHandler('toggleClass', function(data){
		getJqueryObj(data.selector, data.index).toggleClass(data.class)
	})
	python.add.commandHandler('callback', function (data) {
		getJqueryObj(data.selector, data.index)[data.type](function (event) {
			if(this.id==''){
				placeholderIds++;
				this.id = 'PlaceholderId'+placeholderIds;
			}
			py.call(data.name, '#'+this.id);
		});
	});
	
	//requests
	python.add.requestHandler('index', function (data) {
		i = indexedElements.length;
		indexedElements[i] = $(data.selector);
		return i;
	});
	python.add.requestHandler('attribute', function (data) {
		return getJqueryObj(data.selector, data.index).attr(data.property);
	});
	python.add.requestHandler('css', function (data) {
		return getJqueryObj(data.selector, data.index).css(data.property);
	});
	python.add.requestHandler('html', function (data) {
		return getJqueryObj(data.selector, data.index).html();
	});
	python.add.requestHandler('size', function (data) {
		return getJqueryObj(data.selector, data.index).size();
	});
	python.add.requestHandler('value', function (data) {
		return getJqueryObj(data.selector, data.index).value();
	});
	
	function getJqueryObj(selector, index) {
		if (isNaN(index)) {
			return $(selector)
		} else {
			return indexedElements[index]
		}
	}
});




/*
	main API portion
	it is not recomended you make changes to this section unless you know what you are doing
*/


function makePython() {
	platform = require("os").platform;
	root = process.cwd() + '/';

	console.info('nw.js version:' + process.version);
	console.info('platform:' + platform());
	console.info('root directory:' + root);

	if (platform == 'win32') {
		root = root.replaceAll('/', '\\');
		pyDir = pyDir.replaceAll('/', '\\');
	}


	var spawnChild = require('child_process').spawn;

	py = {
		process: spawnChild('python', ['-u', root + pyDir]), //launches python
		onLog: function (data) {
			console.log('python log: ' + data);
		},
		onCommand: {
			
		},
		onRequest:{
			
		},
		call: function (name, eventData) {
			res = '<*>'+name+'<*>'+eventData;
			res = res.replaceAll('\n', ' ');
			res += '\n';
			py.process.stdin.write(res);
		},
		onClose: function (code) {
			console.log('python exited with code ' + code);
		},
		onError: function (data) {
			console.error('python had error:\n' + data);
		},
		add: {
			commandHandler: function (name, callback) {
				py.onCommand[name] = callback;
			},
			requestHandler: function (name, callback) {
				py.onRequest[name] = callback;
			}
		}
	}

	py.process.stdout.on('data', function (msg) {
		msg = msg.toString();
		msg.split('\n').forEach(function (data) {
			if (data.substring(0, 3) == '<*>') {
				data = data.split('<*>');
				py.onCommand[data[1]](JSON.parse(data[2]));
			}else if (data.substring(0, 3) == '<:>'){
				data = data.split('<:>');
				res = py.onRequest[data[1]](JSON.parse(data[2]));
				res = JSON.stringify(res)
				res = '<:>'+data[1]+'<:>'+res;
				res = res.replaceAll('\n', ' ');
				res += '\n';
				py.process.stdin.write(res);
			}else if (data != '') {
				py.onLog(data)
			}
		});
	});

	py.process.stderr.on('data', function (data) {
		data = data.toString();
		py.onError(data);
	});

	py.process.on('close', function (code) {
		py.onClose(code);
	});
	py.kill = function () {
		py.process.kill();
		console.warn('killing Python...')
	}

	return py;
}



/*
	Utils
*/


//Replace last util
String.prototype.reverse = function () {
	return this.split('').reverse().join('');
};

String.prototype.replaceLast = function (what, replacement) {
	return this.reverse().replace(new RegExp(what.reverse()), replacement.reverse()).reverse();
};
//Replace All
String.prototype.replaceAll = function (textToRemove, replacement) {
	return this.split(textToRemove).join(replacement);
}

File: lib/PY/PyQuery.py
from inspect import isfunction
from json import dumps
import os
import sys
#-----------------------------------------------------
#			access using PQ or PyQuery
#If you want to add commands/requests go further down
#-----------------------------------------------------
def makePyQuery():
	def command(name, params = {}):
		msg = dumps(params, separators=(',',':'))
		print('<*>%s<*>%s' % (name, msg))

	def request(name, params = {}):
		msg = dumps(params, separators=(',',':'))
		res = input('<:>%s<:>%s' % (name, msg))
		if(res[0:3]=='<:>'):
			res = res.split('<:>')
			return res[2]

	def startListener():
		while True:
			data = None;
			if sys.version_info >= (3, 0):
				data = input()
			else:
				data = raw_input()

			data = repr(data)
			data = data.replace("'",'')
			data = data.replace('"','')
			if(data[0:3]=='<*>'):
				data = data.split('<*>')
				answerCall(data[1], data[2]);
			else:
				print(data)

	def answerCall(name, data):
		isNameInArray = False
		for callback in callbackList:
			if (callback.get_name() == name):
				callback.run(data)
				isNameInArray=True
				break
		if not isNameInArray:
			command('warn', {'msg':'The %s callback was not found.' % name})

	callbackList = []

	class Callback():
		__name = None
		__function = None
		__selector = None
		__mode = None

		def __init__(self, selector, mode, run=None):
			self.__name = selector + ":" + mode
			self.__selector = selector
			self.__mode = mode
			self.__function = run
			callbackList.append(self)
			if(not isfunction(run)):
				command('warn', {'msg':"The %s callback's function property was not set to a function." % name})

		def run(self, this):
			this=PyQuery(this)
			self.__function(this)

		def set_function(self, function):
			self.__function = function

		def get_name(self):
			return self.__name

		def get_mode(self):
			return self.__mode

		def get_selector(self):
			return self.__selector

	class PyQuery():
		__selector=None
		__index=None
		__callback={}

		def __init__(self, selector):
			self.__selector = selector
			self.__callback = {}
			self.__index = request("index", {'selector':selector})

		def get_selector(self):
			return self.__selector
#--------------------------------------------------------------
#this next section is where you can add more commands/requests
#--------------------------------------------------------------
		def makeCallback(self, mode, runFunc):
			hasMadeCallback = False
			try:
				self.__callback[mode]
				#self.__callback[mode].set_function(runFunc)
			except KeyError:
				self.__callback[mode] = Callback(self.__selector, mode, runFunc)
				hasMadeCallback = True
			if not hasMadeCallback:
				command('warn', {'msg':"The %s's %s callback's function property was already set to a function and can not be set again." % (self.__selector, mode)})

			command('callback', {
					'selector':self.__callback[mode].get_selector(),
					'type':self.__callback[mode].get_mode(),
					'name':self.__callback[mode].get_name(),
					'index':str(self.__index)
				});
			return self
		def append(self, text):
			command('append', {
					'selector':self.__selector,
					'text':text,
					'index':str(self.__index)
				})
			return self
		def css(self, data):
			if(isinstance(data, dict)):
				command('css', {
					'selector':self.__selector,
					'rules':data,
					'index':self.__index
				})
				return self
			else:
				return request('css', {
					'selector':self.__selector,
					'property':data,
					'index':self.__index
				})
		def attr(self, attribute, value=None):
			if value is not None:
				command('attribute', {
						'selector':self.__selector,
						'property':attribute,
						'value':value,
						'index':self.__index
					})
				return self
			else:
				return request('attribute', {
						'selector':self.__selector,
						'property':attribute,
						'index':self.__index
					})
		def html(self, text=None):
			if text is not None:
				command('html', {
						'selector':self.__selector,
						'text':text,
						'index':self.__index
					})
				return self
			else:
				return request('html', {
						'selector':self.__selector,
						'index':self.__index
					})
		def value(self, value=None):
			if value is not None:
				command('value', {
						'selector':self.__selector,
						'value':value,
						'index':self.__index
					})
				return self
			else:
				return request('value', {
						'selector':self.__selector,
						'index':self.__index
					})
		def size(self):
			return request('size', {
					'selector':self.__selector,
					'index':self.__index
				})
		def toggleClass(self, data):
			command('toggleClass', {
					'selector':self.__selector,
					'index':self.__index,
					'class': data
				})

	PyQuery.start = startListener
	PyQuery.command = command
	PyQuery.request = request
	
	return PyQuery

PyQuery = makePyQuery()
PQ = PyQuery		
	